python/demo/gmsh/demo_gmsh.py

```diff
@@ -4,32 +4,82 @@
 #
 # SPDX-License-Identifier:    LGPL-3.0-or-later
 #
-# =====================================
-# Mesh generation using Gmsh and pygmsh
-# =====================================
+# =========================================
+# Mesh generation using the GMSH python API
+# =========================================
 
-import numpy as np
-import pygmsh
-from mpi4py import MPI
+from dolfinx.mesh import create_mesh, create_meshtags
+from dolfinx.io import XDMFFile, ufl_mesh_from_gmsh, read_from_msh, extract_geometry, extract_mesh_topology_and_markers
 
-from dolfinx import cpp
 from dolfinx.cpp.io import perm_gmsh, extract_local_entities
-from dolfinx.io import XDMFFile, ufl_mesh_from_gmsh
-from dolfinx.mesh import create_mesh, create_meshtags
+from dolfinx import cpp
+from mpi4py import MPI
+import gmsh
+import numpy as np
+
+
+# Generating a mesh on one process and write to file
+# ==================================================
+#
+# Generate a mesh on rank 0 and write it to file as msh.
+# Load the mesh and corresponding meshtags as a
+# mesh distributed over all processors
+if MPI.COMM_WORLD.rank == 0:
+
+    gmsh.initialize()
+    model = gmsh.model()
+    model.add("Rectangle")
+    rect_tag = model.occ.addRectangle(0, 0, 0, 1, 1, 1)
+    model.occ.synchronize()
+
+    # Tag the physical volume with tag 7
+    model.addPhysicalGroup(2, [rect_tag], tag=7)
+    # Tag the line [0,0]->[0,1] with tag 11
+    facet_entities = model.occ.getEntities(dim=1)
+    for entity in facet_entities:
+        midpoint = model.occ.getCenterOfMass(dim=1, tag=entity[1])
+    if np.allclose(np.array(midpoint), [0, 0.5, 0]):
+        model.addPhysicalGroup(1, [entity[1]], tag=11)
+    model.mesh.generate(2)
+    gmsh.write("mesh.msh")
+
```

Comment:
Demos seems to be hanging in parallel. Is it possible that a barrier is needed here?

Reply:
Why are we writing to file?

Reply:
To show that we can read from msh.
The issue here was that I only initialized gmsh on one processor, then using it on every processor later on.

Reply:
I think it should be removed, or encapsulated in the demo. DOLFIN supports reading from XDMF. Support for other formats needs to be discussed since it's not clear where one should stop.

Reply:
I disagree, as there is no clear description of how the dolfinx-XDMF format is.
There are plenty of XDMF-files (for instance those created by meshio-convert from msh) that does not work with dolfinx. This makes it possible for any msh mesh to be loaded, and I think that should be highlighted

Reply:
I've removed them from the demo and dolfinx.io, as they would make the demo quite complex if encapsulated.